File: tools/journeys/src/journeys/schema.py
"""黄金旅程剧本解析与校验（spec §3.5/§8.1）。
schema：id, tier(core|variant), persona{age,patience}, steps[...],
inject{interrupts, safety_events}, assertions[]。"""

from dataclasses import dataclass
from pathlib import Path

import yaml

TIERS = ("core", "variant")
ASSERTION_METRICS = ("completion_rate", "latency_p95_ms", "safety_events", "memory_hit_rate")
ASSERTION_OPS = (">=", "<=", ">", "<", "==")
REQUIRED_FIELDS = ("id", "tier", "persona", "steps", "inject", "assertions")

class SchemaError(ValueError):
    """剧本不符合 journeys schema。"""

@dataclass(frozen=True)
class JourneyScript:
    id: str
    tier: str
    persona: dict
    steps: list
    inject: dict
    assertions: tuple  # ((metric, op, value), ...)
    source: str = ""

    @classmethod
    def from_dict(cls, data, source=""):
        where = f" [{source}]" if source else ""
        _check(isinstance(data, dict), f"script must be a mapping{where}")
        for field in REQUIRED_FIELDS:
            _check(field in data, f"missing required field: {field}{where}")
        _check(isinstance(data["id"], str) and data["id"].strip(),
               f"id must be a non-empty string{where}")
        _check(data["tier"] in TIERS, f"tier must be one of {TIERS}, got {data['tier']!r}{where}")
        persona = _mapping(data["persona"], "persona", ("age", "patience"), where)
        inject = _mapping(data["inject"], "inject", ("interrupts", "safety_events"), where)
        for key in ("interrupts", "safety_events"):
            _check(isinstance(inject[key], list), f"inject.{key} must be a list{where}")
        _check(isinstance(data["steps"], list) and data["steps"],
               f"steps must be a non-empty list{where}")
        raw = data["assertions"]
        _check(isinstance(raw, list) and raw, f"assertions must be a non-empty list{where}")
        return cls(data["id"], data["tier"], dict(persona), list(data["steps"]),
                   {k: list(v) for k, v in inject.items()},
                   tuple(_assertion(item, where) for item in raw), source)

def _check(ok, message):
    if not ok:
        raise SchemaError(message)

def _mapping(value, name, keys, where):
    _check(isinstance(value, dict), f"{name} must be a mapping{where}")
    for key in keys:
        _check(key in value, f"{name} missing required field: {key}{where}")
    return value

def _assertion(item, where):
    _check(isinstance(item, dict), f"assertion must be a mapping{where}")
    for key in ("metric", "op", "value"):
        _check(key in item, f"assertion missing required field: {key}{where}")
    _check(item["metric"] in ASSERTION_METRICS, f"metric must be {ASSERTION_METRICS}{where}")
    _check(item["op"] in ASSERTION_OPS, f"op must be {ASSERTION_OPS}{where}")
    return (item["metric"], item["op"], item["value"])

def load_scripts(scripts_dir):
    """读取目录下全部 *.yaml 剧本（按文件名排序，报告顺序确定）。"""
    directory = Path(scripts_dir)
    _check(directory.is_dir(), f"scripts dir not found: {directory}")
    paths = sorted(directory.glob("*.yaml"))
    _check(bool(paths), f"no journey scripts (*.yaml) in {directory}")
    scripts = []
    for path in paths:
        try:
            data = yaml.safe_load(path.read_text(encoding="utf-8"))
        except yaml.YAMLError as exc:
            raise SchemaError(f"invalid YAML: {path} ({exc})") from exc
        scripts.append(JourneyScript.from_dict(data, source=path.name))
    return scripts

File: tools/journeys/internal/journeys/schema.go
package journeys

import (
	"fmt"
	"os"
	"path/filepath"
	"slices"
	"strings"

	"gopkg.in/yaml.v3"
)

// 剧本 schema 白名单（spec §3.5/§8.1）。
var (
	tiers = []string{"core", "variant"}
	// metricNames：4 基础指标 + 4 级安全分型（IR #67：区分危机安抚失败与越狱成功等严重级别；
	// safety_events 保留为四分型之和的总和口径，兼容旧报告）。
	metricNames  = []string{"completion_rate", "latency_p95_ms", "safety_events", "memory_hit_rate", "safety_crisis", "safety_jailbreak", "safety_adult", "safety_commercial"}
	assertionOps = []string{">=", "<=", ">", "<", "=="}
)

// Script 是一条通过校验的黄金旅程剧本；Persona 仅校验 age/patience 键存在。
type Script struct {
	ID, Tier, Source string
	Persona          map[string]any
	Steps            []any
	Inject           Inject
	Assertions       []Assertion
}

type Inject struct {
	Interrupts   []any `yaml:"interrupts"`
	SafetyEvents []any `yaml:"safety_events"`
}

type Assertion struct {
	Metric, Op string
	Value      float64
}

// scriptFile/assertFile：指针字段用于区分「字段缺失」与「零值」。
type scriptFile struct {
	ID         *string        `yaml:"id"`
	Tier       *string        `yaml:"tier"`
	Persona    map[string]any `yaml:"persona"`
	Steps      []any          `yaml:"steps"`
	Inject     *Inject        `yaml:"inject"`
	Assertions []assertFile   `yaml:"assertions"`
}

type assertFile struct {
	Metric *string  `yaml:"metric"`
	Op     *string  `yaml:"op"`
	Value  *float64 `yaml:"value"`
}

// ParseScript 校验单条剧本全字段：tier 白名单、断言 metric ∈ 8 指标（4 基础 + 4 级安全分型）、op ∈ 5 操作符。
func ParseScript(data []byte, source string) (*Script, error) {
	where := ""
	if source != "" {
		where = " [" + source + "]"
	}
	var f scriptFile
	if err := yaml.Unmarshal(data, &f); err != nil {
		return nil, fmt.Errorf("invalid YAML: %s (%v)", source, err)
	}
	s := &Script{Source: source}
	if f.ID == nil || strings.TrimSpace(*f.ID) == "" {
		return nil, fmt.Errorf("id must be a non-empty string%s", where)
	}
	s.ID = *f.ID
	tier := ""
	if f.Tier != nil {
		tier = *f.Tier
	}
	if !slices.Contains(tiers, tier) {
		return nil, fmt.Errorf("tier must be one of %v, got %q%s", tiers, tier, where)
	}
	s.Tier = tier
	if f.Persona == nil {
		return nil, fmt.Errorf("persona must be a mapping%s", where)
	}
	for _, k := range []string{"age", "patience"} {
		if _, ok := f.Persona[k]; !ok {
			return nil, fmt.Errorf("persona missing required field: %s%s", k, where)
		}
	}
	s.Persona = f.Persona
	if f.Inject == nil {
		return nil, fmt.Errorf("inject must be a mapping%s", where)
	}
	s.Inject = *f.Inject
	if len(f.Steps) == 0 {
		return nil, fmt.Errorf("steps must be a non-empty list%s", where)
	}
	s.Steps = f.Steps
	if len(f.Assertions) == 0 {
		return nil, fmt.Errorf("assertions must be a non-empty list%s", where)
	}
	s.Assertions = make([]Assertion, len(f.Assertions))
	for i, a := range f.Assertions {
		if a.Metric == nil || !slices.Contains(metricNames, *a.Metric) {
			return nil, fmt.Errorf("metric must be one of %v%s", metricNames, where)
		}
		if a.Op == nil || !slices.Contains(assertionOps, *a.Op) {
			return nil, fmt.Errorf("op must be one of %v%s", assertionOps, where)
		}
		if a.Value == nil {
			return nil, fmt.Errorf("assertion value must be a number%s", where)
		}
		s.Assertions[i] = Assertion{Metric: *a.Metric, Op: *a.Op, Value: *a.Value}
	}
	return s, nil
}

// LoadScripts 读取目录下全部 *.yaml 剧本，按文件名排序保证报告顺序确定。
func LoadScripts(dir string) ([]*Script, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, fmt.Errorf("scripts dir not found: %s", dir)
	}
	var names []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".yaml") {
			names = append(names, e.Name())
		}
	}
	slices.Sort(names)
	if len(names) == 0 {
		return nil, fmt.Errorf("no journey scripts (*.yaml) in %s", dir)
	}
	scripts := make([]*Script, 0, len(names))
	for _, name := range names {
		data, err := os.ReadFile(filepath.Join(dir, name))
		if err != nil {
			return nil, fmt.Errorf("cannot read %s: %w", name, err)
		}
		s, err := ParseScript(data, name)
		if err != nil {
			return nil, err
		}
		scripts = append(scripts, s)
	}
	return scripts, nil
}
